content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,196 @@
+
+Unlike volumes with `fsGroup`, the devices have no official notion of `deviceGroup`/`deviceUser` that the CRI runtimes (or kubelet)
+would be able to use. We considered using container annotations set by the device plugins (e.g., `io.kubernetes.cri.hostDeviceSupplementalGroup/`) to get custom OCI `config.json` uid/gid values.
+This would have required changes to all existing device plugins which was not ideal.
+
+Instead, we wanted the solution to be *seamless* to end-users without getting the device plugin vendors involved. The selected approach was
+to re-use `runAsUser` and `runAsGroup` values in `config.json` for devices:
+
+```json
+{
+        "type": "c",
+        "path": "/dev/foo",
+        "major": 123,
+        "minor": 4,
+        "fileMode": 438,
+        "uid": <runAsUser>,
+        "gid": <runAsGroup>
+},
+```
+
+With `runc` OCI runtime (in non-[*rootless mode*](docs/tasks/administer-cluster/kubelet-in-userns/)), the device is created (`mknod(2)`) in
+the container name space and the ownership is changed to `runAsUser`/`runAsGroup` using `chmod(2)` (NB: rootless mode and devices is currently not supported).
+Having the ownership updated in the container namespace is justified as the user process is the only one accessing the device. Only `runAsUser`/`runAsGroup`
+are taken into account, and, e.g., the `USER` setting in the container is currently ignored.
+
+While it is likely that the "faulty" deployments (i.e., non-root `securityContext` + devices) do not exist, to be absolutely sure no
+deployments break, we added an opt-in config entry in both containerd and CRI-O to enable the new behavior. The following:
+
+`device_ownership_from_security_context (bool)`
+
+defaults to `false` and must be enabled to use the feature.
+
+## See non-root containers using devices after the fix
+To demonstrate the new behavior, let's use a Data Plane Development Kit (DPDK) application using hardware accelerators, Kubernetes CPU manager, and HugePages as an example. The cluster runs containerd with:
+
+```toml
+[plugins]
+  [plugins."io.containerd.grpc.v1.cri"]
+    device_ownership_from_security_context = true
+```
+
+or CRI-O with:
+
+```toml
+[crio.runtime]
+device_ownership_from_security_context = true
+```
+
+and the Guaranteed QoS Class Pod that runs DPDK's crypto-perf test utility:
+
+```yaml
+...
+metadata:
+  name: qat-dpdk
+spec:
+  securityContext:
+    runAsUser: 1000
+    runAsGroup: 2000
+    fsGroup: 3000
+  containers:
+  - name: crypto-perf
+    image: intel/crypto-perf:devel
+    ...
+    resources:
+      requests:
+        cpu: "3"
+        memory: "128Mi"
+        qat.intel.com/generic: '4'
+        hugepages-2Mi: "128Mi"
+      limits:
+        cpu: "3"
+        memory: "128Mi"
+        qat.intel.com/generic: '4'
+        hugepages-2Mi: "128Mi"
+  ...
+```
+
+`fsGroup` gives a `runAsUser` writable HugePages emptyDir mountpoint. To verity the results, we check the container's environment:
+
+```bash
```

Comment:
IMO separate commands from output and from other commands for better readability.

Reply:
@shannonxtreme any suggestion how to do this? I've now dropped `bash` at least.

Reply:
@shannonxtreme @sftim I have this now fixed. Thanks for the improvement ideas!